src/test/scala/fr/acinq/bitcoin/scalacompat/TaprootSpec.scala

```diff
@@ -0,0 +1,333 @@
+    )
+    Transaction.correctlySpends(tx, inputs, ScriptFlags.STANDARD_SCRIPT_VERIFY_FLAGS)
+  }
+
+  test("use OP_CHECKSIGADD to implement multisig") {
+    val privs = Seq(
+      PrivateKey(ByteVector32.fromValidHex("0101010101010101010101010101010101010101010101010101010101010101")),
+      PrivateKey(ByteVector32.fromValidHex("0101010101010101010101010101010101010101010101010101010101010102")),
+      PrivateKey(ByteVector32.fromValidHex("0101010101010101010101010101010101010101010101010101010101010103"))
+    )
+
+    // we want 2 good signatures out of 3
+    val script = Seq(
+      OP_PUSHDATA(privs(0).xOnlyPublicKey()), OP_CHECKSIG,
+      OP_PUSHDATA(privs(1).xOnlyPublicKey()), OP_CHECKSIGADD,
+      OP_PUSHDATA(privs(2).xOnlyPublicKey()), OP_CHECKSIGADD,
+      OP_2, OP_GREATERTHANOREQUAL
+    )
+
+    // simple script tree with a single element
+    val scriptTree = new ScriptTree.Leaf(new ScriptLeaf(0, Script.write(script), fr.acinq.bitcoin.Script.TAPROOT_LEAF_TAPSCRIPT))
+    val merkleRoot = ScriptTree.hash(scriptTree)
+
+    // we choose a pubkey that does not have a corresponding private key: our funding tx can only be spent through the script path, not the key path
+    val internalPubkey = XonlyPublicKey(PublicKey.fromBin(ByteVector.fromValidHex("0250929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0")))
+    val (tweakedKey, parity) = internalPubkey.outputKey(new fr.acinq.bitcoin.Crypto.TaprootTweak.ScriptTweak(merkleRoot))
+
+    // funding tx sends to our tapscript
+    val fundingTx = Transaction(version = 2, txIn = Nil, txOut = Seq(TxOut(Satoshi(1000000), Seq(OP_1, OP_PUSHDATA(tweakedKey)))), lockTime = 0)
+
+    // create an unsigned transaction
+    val tmp = Transaction(
+      version = 2,
+      txIn = TxIn(OutPoint(fundingTx, 0), ByteVector.empty, fr.acinq.bitcoin.TxIn.SEQUENCE_FINAL) :: Nil,
+      txOut = TxOut(fundingTx.txOut(0).amount - Satoshi(5000), addressToPublicKeyScript(Block.RegtestGenesisBlock.hash, "bcrt1qdtu5cwyngza8hw8s5uk2erlrkh8ceh3msp768v")) :: Nil,
+      lockTime = 0
+    )
+    val hash = hashForSigningSchnorr(tmp, 0, Seq(fundingTx.txOut(0)), SigHash.SIGHASH_DEFAULT, SigVersion.SIGVERSION_TAPSCRIPT, Script.ExecutionData(annex = None, tapleafHash = Some(merkleRoot)))
+
+    // compute all 3 signatures
+    val sigs = privs.map { p => Crypto.signSchnorr(hash, p, fr.acinq.bitcoin.Crypto.SchnorrTweak.NoTweak.INSTANCE) }
+
+    // control is the same for everyone since there are no specific merkle hashes to provide
+    val controlBlock = ByteVector.view((fr.acinq.bitcoin.Script.TAPROOT_LEAF_TAPSCRIPT + (if (parity) 1 else 0)).toByte +: internalPubkey.pub.value.toByteArray)
```

Comment:
Same here, we need a better abstraction for that, this is very low-level details that can be easily hidden from library users.